xwiki-rendering-syntaxes/xwiki-rendering-syntax-annotatedxhtml/src/main/java/org/xwiki/rendering/internal/renderer/xhtml/AnnotatedXHTMLChainingRenderer.java

```diff
@@ -83,4 +87,25 @@ public void endMacroMarker(String name, Map<String, String> parameters, String c
             this.macroRenderer.endRender(getXHTMLWikiPrinter());
         }
     }
+
+    @Override
+    public void beginMetaData(MetaData metadata)
+    {
+
+        Map<String, String> parameters = new HashMap<>();
+
+        for (Map.Entry<String, Object> metadaPair : metadata.getMetaData().entrySet()) {
+            parameters.put(metadaPair.getKey(), metadaPair.getValue().toString());
```

Comment:
Shouldn't you generate ``data-`` attributes? And, if you don't use any namespace then how will you know what is meta data and what is simply HTML attributes added by some other code, when you parse the meta data back into XDOM?

Reply:
> Shouldn't you generate data- attributes?

I thought there was already a processing for non-standard HTML attributes in the printer, so yeah I should :)

> And, if you don't use any namespace then how will you know what is meta data and what is simply HTML attributes added by some other code, when you parse the meta data back into XDOM?

Here, by namespace, you mean a dedicated name, right? Like `data-xwiki-...`, or you mean something more sophisticated such as a DTD? Because on the second case I'm not sure how to implement it.

Reply:
> Here, by namespace, you mean a dedicated name, right? Like ``data-xwiki-``

Yes, maybe not exactly ``data-xwiki-*`` but in any case something to distinguish the attributes that are mapped to XDOM meta data from attributes set by some other code (you never know). Otherwise you end up with all the ``data-`` attributes in the XDOM meta data and this may cause unexpected behavior.

Otherwise, you could write / read only known meta data (instead of everything), but then we would need to update the code when we add a new meta data (which doesn't happen very often, but still).

Reply:
Why not `data-xwiki-*` then? 

> Otherwise, you could write / read only known meta data (instead of everything), but then we would need to update the code when we add a new meta data (which doesn't happen very often, but still).

I could use some reflexivity here: the know metadata should be declared as public String in MetaData.class. WDYT?

Reply:
Hm actually maybe not a good idea, we don't use prefixed attribute to specify that it's a medata so a public string in the class could be anything else.

Reply:
Actually what you imply is that someone could write something like:
```
<div data-syntax="xwiki/2.1" data-custom="foo">
[HTML]
</div>
```
and we should be able to distinguish our own metadata `data-syntax` and a custom one.
But how would we parse such an example? 

```
BeginGroupBlock [[custom]=[foo]]
BeginMetaDataBlock [[syntax]=[xwiki/2.1]]
[PARSED HTML]
EndMetaDataBlock [[syntax]=[xwiki/2.1]]
EndGroupBlock [[custom]=[foo]]
```

but then when rendering this again we would obtain:
```
<div data-custom="foo">
<div data-syntax="xwiki/2.1">
[HTML]
</div>
</div>
```

So I'd say on the contrary that the div or span we create for the metadata use a dedicated class or id which specify that it's an element created to support XWiki metadata only. Any other custom attribute should be placed in another element. WDYT?

Reply:
My suggestion was to ignore the custom attributes. For instance, if you parse

    <div class="wikimodel-emptyline" data-custom="foo">

the custom attribute is lost AFAIK.

Reply:
OK I misunderstood but now we're on the same page :)